File: defaults/kdj.js
function* kdj(C){
    //指标定义
    C.DEFINE({
        type: "SUB",
        cname: "KDJ",
        state: "KLINE",
    });
    //参数
    let n = C.PARAM(3, "N");
    let m1 = C.PARAM(5, "M1");
    let m2 = C.PARAM(5, "M2");
    //输入序列
    let close = C.SERIAL("CLOSE");
    let high = C.SERIAL("HIGH");
    let low = C.SERIAL("LOW");
    //输出序列
    let k = C.OUTS("LINE", "k", {color: RED});
    let d = C.OUTS("LINE", "d", {color: GREEN});
    let j = C.OUTS("LINE", "j", {color: YELLOW});
    //临时序列
    let rsv = [];
    //计算
    while(true) {
        let i = yield;
        let hv = HIGHEST(i, high, n);
        let lv = LOWEST(i, low, n);
        rsv[i] = (hv == lv) ? 0 : (close[i] - lv) / (hv - lv) * 100;
        k[i] = SMA(i, rsv, m1, 1, k);
        d[i] = SMA(i, k, m2, 1, d);
        j[i] = 3*k[i] -2*d[i];
    }
}

File: script/cases/kdj.py
#!/usr/bin/env python
#coding=utf-8

case = {
    "id": "KDJ",
    "cname": "KDJ",
    "type": "SUB",
    "src": """
//该模型仅仅用来示范如何根据指标编写简单的模型
//用户需要根据自己交易经验，进行修改后再实际应用!!!
// //后为文字说明，编写模型时不用写出
RSV:=(CLOSE-LLV(LOW,N))/(HHV(HIGH,N)-LLV(LOW,N))*100;//收盘价与N周期最低值做差，N周期最高值与N周期最低值做差，两差之间做比值。
K:SMA(RSV,M1,1);//RSV的移动平均值
D:SMA(K,M2,1);//K的移动平均值
J:3*K-2*D;
BACKGROUNDSTYLE(1);
""",
    "params": [
        ("N", 1, 100, 9),
        ("M1", 2, 100, 3),
        ("M2", 2, 100, 3),
    ],
    "expected": """
    """,
}
